test: scene transitioning sample and global game state management

NoelStephensUnity opened the pull request:
This is the first pass scene transitioning and of a global game state management sample.
It provides an "out of the box" solution that a user can use as working framework to expand upon.
Scenes are linked to both game and MLAPI states (via a custom UI property field control) that provide clean scene transitioning with the ability to join during an in progress game session without any hassle or "soft synch" issues.
**There are two networked state machines:**
- The Global Game State Machine
- The In-Game State Machine

**The default scenes that come with this sample are:**
MLAPIBootStrap: Contains the NetworkingManager, UNet Transport, and GlobalGameState components
GameIntro: Nothing but a place holder scene that users can populate with additional intro spash screens or videos
MainMenu: Very simple main menu that provides to options: "Start as Host" or "Start as Client"
GameLobby: It has a simple interface (for now), but has the inner-workings to build a full fledge lobby upon.  This also provides several NetworkedVar use case scenarios and provides users with a way to "hide players until they enter an In-Game scene".
InGame, InGame2, InGame3:  The "multiplayer game scenes" in this sample, this version the players move on their own randomly.  The server can pause the game, which demonstrates how to pause an in-session game using MLAPI and Unity (at least one way to do it).  If the host-server exits while clients are still present, then a count down will commence.  Upon the counter reaching zero (0), all clients will be removed from the multiplayer session.  Upon all clients being removed from the multiplayer session, the host-server will then exit.

The sample also provides the user with the ability to launch a scene without having to have a networking manager in the scene itself, and GlobalGameState handles boot strapping MLAPI and then loading the current scene opened within the editor.  
In order for this to work properly you must have registered:
- All scenes that require or MLAPI or will transition into a scene that requires MLAPI with the NetworkingManager
- All scenes for the entire game within the Build Settings -> Scenes In Build 
- All scenes with GlobalGameState which also requires linking the MLAPI state (if any) and Global Game State to the scene.

The visual goal was to minimize the assets required to simplify the exporting process.  All scripts and prefabs I tried to keep completely contained within those folders to make this process as straight forward as possible for the user.

_The PlayerControl.cs is still WIP._

0xFA11 committed:
implement NetworkUpdateLoop

0xFA11 committed:
merge branch 'develop' into 'experimental/network-update-loop'

0xFA11 committed:
merge branch 'develop' into 'experimental/network-update-loop'

0xFA11 committed:
merge branch 'develop' into 'experimental/network-update-loop'

0xFA11 committed:
remove AdvanceFrame() and FrameCount

0xFA11 committed:
update player loop injection points and INetworkUpdateSystem API

0xFA11 committed:
add xmldoc and notes

0xFA11 committed:
add NetworkUpdateLoopTests

0xFA11 committed:
implement NetworkUpdateLoop plain and mixed tests

0xFA11 committed:
update xmldoc

0xFA11 committed:
small meta fix

0xFA11 committed:
replace old network update loop with new RFC-backed network update loop

0xFA11 committed:
replace WaitForEndOfFrame with WaitUntil in tests

0xFA11 committed:
comment 'default' for NetworkUpdateStage.Update enum value

0xFA11 committed:
optimization & refactoring pass

0xFA11 committed:
Merge branch 'develop' into feature/network-update-loop

0xFA11 committed:
dictionary lookups instead of index access to arrays

0xFA11 committed:
optimize array allocation

0xFA11 committed:
Merge branch 'develop' into feature/network-update-loop

0xFA11 committed:
Merge branch 'develop' into feature/network-update-loop

NoelStephensUnity committed:
Fix:  PostProcessing and Resetting NetworkingManager transport

This fixes two issues:
It prevents the postprocessing method ProcessScene from being run during scene transitions while in PlayMode within the editor.

This fixes the issue where the transport is no longer being polled for events upon shutting down the server,host, or client.

NoelStephensUnity committed:
fix: more robust fix for postprocessing bug

This prevents this post processing code from being executed when in PlayMode (editor and stand alone build)

NoelStephensUnity committed:
Merge branch 'develop' into experimental/scenemanagement

NoelStephensUnity committed:
refactor: SceneSwitchProgress

This ended up not being needed in the end.
Reverting back to the original file.

NoelStephensUnity committed:
refactor: removing commented code and merge artifacts

Removing the commented out OnDisable function (the containing code was moved into the Shutdown method) and removed the Awake method (the containing code was moved into the Init method).

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt512

NoelStephensUnity committed:
refactor: merge artifacts

This hadn't fully registered the merge.

NoelStephensUnity committed:
First pass adjustments

Just getting the footprint ready and made some adjustments to the ground and wall materials.

NoelStephensUnity committed:
test: initial setup for the scene transistion sample

WIP files for the scene transition samples.

NoelStephensUnity committed:
syntax: comments

Refactoring some comments

NoelStephensUnity committed:
WIP: State to Scene Property Work

This update is still WIP, but it will allow the user to associate scenes with the game state in which they should be loaded.

NoelStephensUnity committed:
refactor: Scene to State links

Cleaned up the code a bit, setup the primary skeleton scenes, and have tested up to launching into the lobby scene.  Still requires some additional polish, creating a new player prefab, putting together a simple lobby interface, and some simple game mechanics in the InGame scene (i.e. players move a 2d object on screen).
However, this is the core framework for this sample.

NoelStephensUnity committed:
refactor: Auto-Bootstrap MLAPIBootStrap

Some adjustments to the scenes as well as added the ability to auto load the boot strap when you are entering PlayMode from within the editor but are not on the default MLAPIBootStrap scene.

NoelStephensUnity committed:
refactor: Scene to Scene Name

This handles extracting the scene name from the Scene object for runtime usage since scene objects don't make it into built packages.

NoelStephensUnity committed:
refactor: switched to SceneAsset for in editor assignment

Decided to switch the UnityEngine.Object to UnityEditor.SceneAsset in order to assure nothing but a scene is assigned to the scene field.

NoelStephensUnity committed:
fix: Remove SceneAsset from build

Forgot to wrap the SceneAsset in #if (UNITY_EDITOR) to exclude it from any builds.

NoelStephensUnity committed:
update: Lobby to InGame back to MainMenu

This update (still WIP) now transitions between in session scenes and resets MLAPI properly.

NoelStephensUnity committed:
fix: fixes client not having all players upon first connecting

This update fixes the issue where the client would enter the lobby but not have all players listed until they clicked the ready button.  On the client side, I am pumping the server by notifying that I am "here" (even though the server knows this) but not ready.  The server, upon receiving the first status message, will apply the status and then send a notification out to all clients on the status of players in the lobby.

NoelStephensUnity committed:
Update: Cameras, Rigid bodies, and more!

This is still WIP, but hiding players with some default items already in the samples is working, pausing appears to be working other than the SyncTransform continuing to update even when the rigid body's constraint is all.
Other than a better playerObject (tomorrow), this is a good point to get feedback.

NoelStephensUnity committed:
fix and format:  Client camera and CR

Fixing some weird LFCR issues that started happening recently.
Fixed client camera not being synchronized when joining a game already in progress.
Fixed client not being able to exit.
Added a flag to the in game manager to tell it whether it should exit upon being the last player or not.

NoelStephensUnity committed:
Refactor: player camera

A better way to handle the player camera assignment.
Minor text adjustment for host-server side commands text

NoelStephensUnity committed:
refactor: Players movement, collision, colors, and more

This update is close to the final version with potentially some additions to the graphics or the like in the next week.  The core framework for all of the systems are in place, players can join via lobby or join during game play without soft synch issues or issues with initializing positions.

Did some clean up of the comments, scenes (added exit button for lobby), etc.

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
fix: accessing RandomPlayerMover after client exited

This fixes the issue with trying to still access a client after running through the exiting timer count down.
This also increases the exiting font size and brightening the color a tad bit.

NoelStephensUnity committed:
refactor and updated comments

Removed some code not needed in the LobbyControl.
Added additional comments to some methods that were overlooked.
Fixed some syntax related issue in all edited files.
Removed some unused properties.

NoelStephensUnity committed:
refactor: merge artifact

One too many calls to "UnregisterAllNetworkUpdates", had to be something from the develop merge.

NoelStephensUnity committed:
style: minor spacing

Just adding some minor spaces after if statements.

NoelStephensUnity committed:
style: adjustments for our standard

Found a few old "bad habits" that I needed to fix.

NoelStephensUnity committed:
fix: re-linking text to lobby and populating GlobalGameState

Due to some style-standards adjustments, some assets had to be re-linked due to the changes in their property field name.  This fixes that issue.

NoelStephensUnity committed:
refactor and style

Removed an unused OnDestroy method and added use case scenarios for the RPCs in the LobbyControl

NoelStephensUnity committed:
style

Found several places were there was no space after the If statement.
Removed unused namespaces.

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
refactor: minor adjustment to ingame

Setting the default editor launching mode within the InGame scene to be host mode.

NoelStephensUnity committed:
refactor: moving all related assets under one folder

After seeing how nice and clean Faith's chick sample was, I decided to follow his pattern of keeping all pertinent assets under a single folder.
This commit is just moving those assets to a GlobalGameState folder.

NoelStephensUnity committed:
Refactor: Transition to Multiple InSession Scenes.

This provides an example of how to transition to multiple in game scenes.

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
style: fixed level name in-game

Changed the font to white so it is easier to read.

NoelStephensUnity committed:
style: Standards

Replacing public camel case properties with public pascal case properties.

LukeStampfli commented on testproject/Assets/GlobalGameState/Scripts/InGameManager.cs:
```diff
@@ -0,0 +1,442 @@
+    /// <param name="shouldFreeze">should we freeze everyone or un-freeze them?</param>
+    void FreezeAndShowPlayers(bool shouldFreeze = false)
+    {
+        NetworkedObject[] NetoworkedObjects = GameObject.FindObjectsOfType<NetworkedObject>();
```
SpawnManager.SpawnedObjects could be used instead to show best practices/ better performance maybe?

NoelStephensUnity replied:
Very nice catch Luke!
Yeah, that is a much better approach a provides yet another use case scenario of MLAPI features.
As opposed to parsing through all GameObjects looking for the existing NetworkedObjects, just use the existing SpawnedObjects within the SpawnManager.  (your suggestion has been applied!)
TY!  :)

NoelStephensUnity committed:
refactor: better networked object parsing

Suggested improvement and better MLAPI use case by Luke Stampfli:
Use the SpawnManager's SpawnedObjects list to provide a better use case scenario/example of using MLAPI's current features.  As opposed to searching all GameObjects for a specific type, just use the SpawnManager's already existing list of spawned network object.

LukeStampfli commented on testproject/Assets/GlobalGameState/Scripts/InGameManager.cs:
```diff
@@ -0,0 +1,444 @@
+    /// ServerCommandInputUpdate
+    /// This is where you can add commands or detect key presses to perform server/host operations
+    /// </summary>
+    void ServerCommandInputUpdate()
```
Command might confuse some people because UNet/Mirror use Command to refer to ServerRPCs

LukeStampfli reviewed (approved):
This is a really neat example to show users a way they can do scene transitions and just over all architecture with a state machine in MLAPI. I think this will be really useful to a lot of people.

There is a companion document which goes into the details of how this is structured and explains details. I think it would be really useful if we could transfer some of that information into a readme and add it to the project.

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
refactor: name changes

Changing all of the names to the new naming convention for all things who's names changed.

NoelStephensUnity commented:
Please not merge this yet.  I need to refactor some aspects to it.

NoelStephensUnity committed:
fix: remove singleton references

This removes the NetworkManager.Singleton dependencies from within the RpcQueueContainer and RpcQueueProcessor classes.
This further strengthens the RPC Queue system as a whole, will allow for multiple instances if RpcQueueContainer, removes a limitation I set in NetworkWriter.ReadAndWrite,  and provides unit testing capabilities required for new RPC Queue unit Tests.

NoelStephensUnity committed:
test: rpc queue tests

The new rpc queue tests without all of the extra files associated with them.

NoelStephensUnity committed:
test: ClientRpcTesting

Added some additional prefabs to make creating a new scene simpler.
Added a ManualTests folder where manual tests can live.
Added ClientRpcTesting scene that tests sending to one client, multiple clients, and all clients in many different combinations.

NoelStephensUnity committed:
Merge branch 'develop' into test/multiclienttests

NoelStephensUnity committed:
Merge branch 'develop' into test/multiclienttests

NoelStephensUnity committed:
refactor: removed empty GameObject

Noticed there was an empty GameObject in the testing scene, so I removed it.

NoelStephensUnity committed:
Merge branch 'develop' into test/multiclienttests

NoelStephensUnity committed:
refactor: condense, better comments, minor polish

Condensed the initialization process.
Adjusted the comments to better match the user story (testing and what is being tested).
Adjusted-polished up some UI related items like the visual name or concept relating to a specific counter, etc.

NoelStephensUnity committed:
style: standards

Minor LF adjustment and permissions declaration for m_GlobalCounterOffset property

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
Merge branch 'test/multiclienttests' into test/scenetransitioning

NoelStephensUnity committed:
refactor: several updates and synch with develop branch

This contains several refactoring related changes to get the global game state and scene transitioning sample working again.
Migrated all NetworkVars out of the GlobalGameState and into the GlobalGameStateComponent that is instantiated upon the host entering the lobby.  This dynamically spawned prefab persists until the multiplayer session is ended.
Several style related updates and adjustments to comments.

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
style and standards

Fixed some styling and standards related issues.

NoelStephensUnity committed:
style

bool to var (standards)

NoelStephensUnity committed:
style: standards

Some more standards related changes.

NoelStephensUnity committed:
style: standard

added private in front of IsAlive

NoelStephensUnity committed:
refactor: remove legacy scene

Removed legacy scene and associated meta file.

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

NoelStephensUnity committed:
Merge branch 'develop' into test/scenetransitioning

TwoTenPvP commented:
@NoelStephensUnity Can this be closed?

NoelStephensUnity commented:
Yeah... go ahead... it is way out of date.  It was a good thought at the time.  
Keep the branch please.  :)
